### PR TITLE
Fix all build warnings and treat all future warnings as an error

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1859

Regression? Last working version:

## Description

Fix all existing warnings in our build and enable `MSBuildTreatWarningsAsErrors` so that no new warnings are introduced.

* Set `VisualStudioVersion` to 17.0 otherwise we get a warning about using VS 2019 to build .NET 6 even though we are using a newer version of .NET SDK
* Remove `VisualStudioVersion` global property from non-Windows builds since its set by our own build logic
* Remove redundant imports to `$(MicroBuildDirectory)Microsoft.VisualStudioEng.MicroBuild.Core.props`
* Remove missing files in `src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore`
* Use an older version of `System.Configuration.ConfigurationManager` in `test/NuGet.Core.Tests/NuGet.Packaging.Test/NuGet.Packaging.Test.csproj` when targeting .NET Core 3 since that version of the runtime uses an older version and we get resolve assembly reference warnings otherwise
* Manually consume the assets of `Microsoft.Test.Apex.VisualStudio` since it contains a bunch of DLLs in the `lib` folder that generate compilation warnings because they are not built against `AnyCPU`

There are still a bunch of warnings emitted by other parts of our build which I can not seem to fix:

image

These are coming from DartLab execution and I've found at least one related issue:
-  https://github.com/microsoft/vstest/issues/3208

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: build/packages.targets
Comment (nkolev92) [inline build/packages.targets]: We try to sort the packages alphabetically.

File: build/packages.targets
Comment (nkolev92) [inline build/packages.targets]: I'm not sure I understand why we need the framework check? 
That's the package version we're setting anyways.

File: test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj
Comment (nkolev92) [inline test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj:33]: I'm not necessarily excited about this change. 

Are the compiler warnings really worth referencing the assemblies in a custom way? Can we just no warn that instead?

**Review by nkolev92 — COMMENTED** [review]
Happy to see we're resolving CI warnings :)

File: build/packages.targets
Comment (jeffkl) [reply build/packages.targets]: I only added an top-level package reference for netcoreapp3.1 here: https://github.com/NuGet/NuGet.Client/pull/4799/files#diff-907eee8ef7abe34ee61453861c66e151a3593939832f72110dc2052c1b8f7f2eR22-R29

So I figured I'd add the target framework condition for the Update as well just to be a good citizen.  Should I remove the condition?

Comment (nkolev92) [reply build/packages.targets]: It'd be great if we didn't have conditions. 

We're already using 4.4.0 of the ProtectedData package, https://github.com/NuGet/NuGet.Client/blob/a56c42dbdd901c0c611eece400ce8565fffdfdf7/build/packages.targets#L61, do we even need this addition?

I might be missing something.

File: test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj
Comment (jeffkl) [reply test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj:33]: > I'm not necessarily excited about this change.

This one definitely feels like the least worst option 😄 

Here is what I considered:

1. The package has a terrible layout with an EXE in the `lib` folder and all of its runtime dependencies next to it
  image
2. The package contains a `.targets` file that copies everything from the `tools` directory
  ```xml
<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <None Include="$(MSBuildThisFileDirectory)..\tools\*">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Visible>false</Visible>
    </None>
  </ItemGroup>
</Project>
  ```
3.  The `tools` folder contains actual runtime dependencies
  image

At first I checked if a newer version of the package had a better layout, no dice.

Then I looked at suppressing the warning but the compiler warning is generic, the reference assembly is a different CPU, so suppressing the warning would apply to the entire project and all of its references.  If you know of a way to suppress this warning for a single assembly I'll gladly use that instead.  I would NoWarn the PackageReference instead but that doesn't work for this scenario.

I could write a custom target that _removes_ the references that cause the warning right after `ResolveAssemblyReferences` but I felt that would be harder to understand.

So I settled on manual consumption of the compile-time assets.  I'm open to other suggestions!

Comment (nkolev92) [reply test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj:33]: I'll let others comment. 

Can we ask the owners to fix this, given that it's internal? 

We don't build apex tests locally often enough, so I'm not too worried about it as long it works.

Comment (aortiz-msft) [reply test/NuGet.Tests.Apex/NuGet.OptProf/NuGet.OptProf.csproj:33]: +1. Let's email the owners to fix the package. In the meantime, I think we can take these changes.

**Review by aortiz-msft — DISMISSED** [review]

File: build/packages.targets
Comment (jeffkl) [reply build/packages.targets]: Updated, seems to be working...

**Review by nkolev92 — DISMISSED** [review]
Raise all the errors 🥳 🥳 🥳 🥳 🥳 🥳

**Comment by jeffkl** [comment]
@NuGet/nuget-client be aware that once this change is merged, we will not be able to introduce new warnings to the build.  If you hit a warning that you cannot figure out, feel free to ping me and I'll help you determine how to fix it.

**Review by zivkan — DISMISSED** [review]
I have mixed feelings about warnings as errors. Firstly, is destroys the semantic meaning between warnings and errors. It's basically a signal that people using it thinks that warnings shouldn't exist and there should only be errors.

More pragmatic though is that when build agents update to newer versions of VS and msbuild, if we have to service an older version of Nuget, the build might start failing because new warnings were added, and now servicing the old branch becomes more difficult & risky because we have to fix the warnings in addition. We shot ourselves in the foot in exactly this way with the duplicate PackageReference warning. It caused our supported branches to start failing because something out of our control (the software installed on the CI agent) changed, not because our source code changed.  This PR increases that risk.

On the other hand, warnings are supposed to inform us of potential problems, so minimising risk of adding new potential problems is good 🤷

**Comment by jeffkl** [comment]
> More pragmatic though is that when build agents update to newer versions of VS and msbuild, if we have to service an older version of Nuget, the build might start failing because new warnings were added, and now servicing the old branch becomes more difficult & risky because we have to fix the warnings in addition. We shot ourselves in the foot in exactly this way with the duplicate PackageReference warning. It caused our supported branches to start failing because something out of our control (the software installed on the CI agent) changed, not because our source code changed. This PR increases that risk.

@zivkan I agree and was thinking the exact same thing which is why I haven't merged.  My intent is to keep anyone from introducing new warnings so a warning should block a local build and a PR build.  But I would hate to see an official build blocked because of some warning.  Our official build ignores test failures, it might as well ignore new or intermittent warnings.

```xml
<MSBuildTreatWarningsAsErrors Condition="'$(isOfficialBuild)' != 'true'">true</MSBuildTreatWarningsAsErrors>
```

As for new warnings in the SDK, we currently lock to a specific version of .NET SDK so when we update to a newer version, the commit would also address new warnings.  I am concerned about newer versions of Visual Studio/MSBuild breaking our builds so we should probably lock our builds to specific version of them as well with something like `windows-2022`.  Hopefully a new warning will only be on a major version.

I'm also completely open to disabling this for all hosted builds if it ends up blocking us like crazy.

**Comment by nkolev92** [comment]
> As for new warnings in the SDK, we currently lock to a specific version of .NET SDK so when we update to a newer version, the commit would also address new warnings. 

Is this actually a thing? We just follow the agent upgrades afaik.

**Comment by jeffkl** [comment]
> > As for new warnings in the SDK, we currently lock to a specific version of .NET SDK so when we update to a newer version, the commit would also address new warnings.
> 
> Is this actually a thing? We just follow the agent upgrades afaik.

Yeah we set the version of the SDK to install on agents: 

https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L46

**Comment by nkolev92** [comment]
I think that specifies the SDK we use in our integration tests, not the SDK we build the project with.

I'll defer to someone that's done a lot more CI work recently though, cc @zivkan

**Comment by zivkan** [comment]
Yes, that should be correct. We should use the "system" dotnet cli for building, and we should use the system dotnet cli to run the test assemblies (`dotnet test`).  But our `test/NuGet.Core.FuncTests/dotnet.integration.tests` project will copy the SDK specified in config.props, overwrite the nuget.*.dll files, and then run tests on it. In theory, that should be the only place that specific version of the .NET SDK is used.

**Review by zivkan — APPROVED** [review]

**Comment by ghost** [comment]
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

**Comment by jeffkl** [comment]
I'm merging this for now, if builds start failing we can at least disable treating all warnings as errors and I can just fix warnings as they come up.

**Pull request merged**
